rules/detect-trojan-source.js

```diff
@@ -0,0 +1,62 @@
+/**
+ * Detect trojan source attacks that employ unicode bidi attacks to inject malicious code
+ * @author Liran Tal
+ */
+
+'use strict';
+
+//-----------------------------------------------------------------------------
+// Requirements
+//-----------------------------------------------------------------------------
+
+const { hasTrojanSource } = require('anti-trojan-source');
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'error',
+    docs: {
+      description: 'Detect trojan source attacks that employ unicode bidi attacks to inject malicious code.',
+      category: 'Possible Security Vulnerability',
+      recommended: true,
+      url: 'https://github.com/nodesecurity/eslint-plugin-security#detect-trojan-source',
+    },
+  },
+  create: function (context) {
+    return {
+      onCodePathStart: function (codePath, node) {
+        // at the start of analyzing a code path
+        if (node.tokens && Array.isArray(node.tokens)) {
+          node.tokens.forEach((tokenObject) => {
+            if (tokenObject.value && hasTrojanSource({ sourceText: tokenObject.value })) {
+              context.report({
+                node: node,
```

Comment:
In order to correctly report the location of the character, we need to manually pass in the `loc`, otherwise every report will have a location of line 1, column 0. See https://eslint.org/docs/latest/developer-guide/working-with-rules#contextreport.

Here's an example:
https://github.com/eslint/eslint/blob/main/lib/rules/no-tabs.js#L59-L72

You should also adds checks into the tests to verify that the location is correct.

Because these characters are invisible, having the correct location is really important.

Reply:
We spent a bit of time working on this one and we realised that the original `detect-trojan-source` by @lirantal does not give you granularity on where exactly every single malicious unicode character is. On the contrary it just gives you a boolen telling if the given code token or comment contains any of such characters.

Because of this, the reporting is a bit vague and it just give users a reference about the affected token.

We changed the code slightly to report the location of the token (and updated tests accordingly).

We could improve this further by changing the `hasTrojanSource` function to return a list of malicious unicode characters and their offset in the token.

This would make the scan a bit more expensive though, so we are not too sure whether it is a worthwhile approach or not.

There's a third option that might be a decent tradeoff between accuracy and performance. Rather than returning a boolean `hasTrojanSource` could return the offset of the first occurrence of a malicious unicode character. This way, we will still have one single error per token, but we could at least point the user to the first occurrence of the hidden character. Once the user fixes that, it should be prompter to the next character (if there's more than one per token).

@nzakas what do you think? Do you have any preference?

Reply:
I like that third option which is sorta _fail fast_ as a good balance for quickly running through the code, if it's costly. Also, would it make any sense to push these updates to the `detect-trojan-source` package, or is that at all not used here anyway?

Reply:
Happy to do a PR with any relevant change to `detect-trojan-source` back to the original repo (as of now we did not change any code coming from there, just copied verbatim the `hasTrojanSource` and pasted it into the codebase here).

Thanks a lot for the suggestion of option 3. For now my favourite too :)

Reply:
@lmammino I think you may be overthinking this a bit. In general, people don't like rules that gradually reveal errors -- they want to see them all up front. If it were me, I would expect warnings about every instance of a bidi character in the source code so I'd know how many issues we are talking about.

A good example to follow is the ESLint core rule [`no-tabs`](https://github.com/eslint/eslint/blob/main/lib/rules/no-tabs.js), which does essentially the same thing, only it's looking for tab characters instead of bidi characters.

Reply:
Solved using a similar approach to `no-tabs`